Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,1422 @@
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+        "Client is closed, cannot send a message."
+      )
+    }
+  }
+
+  /// Returns the client's next request to the server.
+  /// - Returns: The request to be made to the server.
+  private mutating func clientNextOutboundMessage() throws -> OnNextOutboundMessage {
+    switch self.state {
+    case .clientIdleServerIdle:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError("Client is not open yet.")
+    case .clientOpenServerIdle(var state):
+      let request = try state.framer.next(compressor: state.compressor)
+      self.state = .clientOpenServerIdle(state)
+      return request != nil ? .sendMessage(request!) : .awaitMoreMessages
+    case .clientOpenServerOpen(var state):
+      let request = try state.framer.next(compressor: state.compressor)
+      self.state = .clientOpenServerOpen(state)
+      return request != nil ? .sendMessage(request!) : .awaitMoreMessages
+    case .clientClosedServerIdle(var state):
+      let request = try state.framer.next(compressor: state.compressor)
+      self.state = .clientClosedServerIdle(state)
+      if let request {
+        return .sendMessage(request)
+      } else {
+        // If the client is closed and there is no message to be sent, then we
+        // are done sending messages, as we cannot call send(message:) anymore.
+        // There are no more messages to be sent, so we can end the compressor.
+        state.compressor?.end()
+        return .noMoreMessages
+      }
+    case .clientClosedServerOpen(var state):
+      let request = try state.framer.next(compressor: state.compressor)
+      self.state = .clientClosedServerOpen(state)
+      if let request {
+        return .sendMessage(request)
+      } else {
+        // If the client is closed and there is no message to be sent, then we
+        // are done sending messages, as we cannot call send(message:) anymore.
+        // There are no more messages to be sent, so we can end the compressor.
+        state.compressor?.end()
+        return .noMoreMessages
+      }
+    case .clientOpenServerClosed(let state):
+      // No point in sending any more requests if the server is closed:
+      // we end the compressor.
+      state.compressor?.end()
+      return .noMoreMessages
+    case .clientClosedServerClosed(let state):
+      // No point in sending any more requests if the server is closed:
+      // we end the compressor.
+      state.compressor?.end()
+      return .noMoreMessages
+    }
+  }
+
+  private mutating func clientReceive(
+    metadata: HPACKHeaders,
+    endStream: Bool
+  ) throws -> OnMetadataReceived {
+    switch self.state {
+    case .clientOpenServerIdle(let state):
+      guard metadata.grpcStatus != nil || metadata.status == "200" else {
+        let httpStatusCode = metadata.status
+          .flatMap { Int($0) }
+          .map { HTTPResponseStatus(statusCode: $0) }
```

Comment:
NIT: Any reason to not write this in a single line?

Reply:
You mean why not `.flatMap { HTTPResponseStatus(statusCode: Int($0)) }`?
`Int($0)` can return `nil` so this way I avoid having to explicitly unwrap the optional.